compiler/hie-core/src/Development/IDE/LSP/LanguageServer.hs

```diff
@@ -12,144 +11,102 @@ module Development.IDE.LSP.LanguageServer
+runLanguageServer getIdeState = do
+    -- Move stdout to another file descriptor and duplicate stderr
+    -- to stdout. This guards against stray prints from corrupting the JSON-RPC
+    -- message stream.
+    newStdout <- hDuplicate stdout
+    stderr `hDuplicateTo` stdout
+
+    -- Print out a single space to assert that the above redirection works.
+    -- This is interleaved with the logger, hence we just print a space here in
+    -- order not to mess up the output too much. Verified that this breaks
+    -- the language server tests without the redirection.
+    putStr " " >> hFlush stdout
+
+    clientMsgChan :: TChan AddItem <- newTChanIO
+    -- These barriers are signaled when the threads reading from these chans exit.
+    -- This should not happen but if it does, we will make sure that the whole server
+    -- dies and can be restarted instead of losing threads silently.
+    clientMsgBarrier <- newBarrier
+
+    let runHandler = RunHandler
+            (\wrap f -> Just $ \r -> atomically $ writeTChan clientMsgChan $ AddResponse r wrap f)
+            (\f -> Just $ \r -> atomically $ writeTChan clientMsgChan $ AddNotification r f)
+    handlers <- mergeHandlers [LS.Definition.addGotoDefinition, LS.Hover.addOnHover, addNotifications, addIgnored] runHandler def
+
+    void $ waitAnyCancel =<< traverse async
+        [ void $ LSP.runWithHandles
+            stdin
+            newStdout
+            ( const $ Right ()
+            , handleInit (signalBarrier clientMsgBarrier ()) clientMsgChan
+            )
+            handlers
+            options
+            Nothing
+        , void $ waitBarrier clientMsgBarrier
+        ]
+    where
+        handleInit :: IO () -> TChan AddItem -> LSP.LspFuncs () -> IO (Maybe err)
+        handleInit exitClientMsg clientMsgChan lspFuncs@LSP.LspFuncs{..} = do
+            ide <- getIdeState sendFunc (makeLSPVFSHandle lspFuncs)
+            _ <- flip forkFinally (const exitClientMsg) $ forever $ do
+                msg <- atomically $ readTChan clientMsgChan
+                case msg of
+                    AddNotification NotificationMessage{_params} act -> act ide _params
+                    AddResponse RequestMessage{_id, _params} wrap act -> do
+                        res <- act ide _params
+                        sendFunc $ wrap $ ResponseMessage "2.0" (responseId _id) (Just res) Nothing
+            pure Nothing
+
+
+-- | Things that come up regularly, but we don't deal with
+addIgnored :: RunHandler -> LSP.Handlers -> IO LSP.Handlers
+addIgnored _ x = return x
+    {LSP.cancelNotificationHandler = none
+    ,LSP.initializedHandler = none
+    ,LSP.codeLensHandler = none
+    }
+    where none = Just $ const $ return ()
+
+
+data AddItem
+    = forall m req resp . AddResponse (RequestMessage m req resp) (ResponseMessage resp -> FromServerMessage) (IdeState -> req -> IO resp)
+    | forall m req . AddNotification (NotificationMessage m req) (IdeState -> req -> IO ())
```

Comment:
Not sure in what sense it could be IO. You could lift the NotificationMessage outside this function, but then my concern is people might pattern match and fail outside a tracked context. My plan is to add exception handlers at some point.

Reply:
Sounds reasonable, :+1: on using existentials for this then.